### PR TITLE
[Curp] Client redirect and test utils

### PR DESCRIPTION
**Pull request by markcty**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,20 +23,21 @@ jobs:
...
+          token: ${{ secrets.GITHUB_TOKEN }}
 
   check:
     name: Check
     runs-on: ubuntu-latest
     steps:
-      - run: sudo apt install -y libprotobuf-dev protobuf-compiler
       - uses: actions/checkout@v2
+      - run: sudo apt install -y cmake g++
+      - run: wget https://github.com/protocolbuffers/protobuf/releases/download/v21.10/protoc-21.10-linux-x86_64.zip && unzip protoc-21.10-linux-x86_64.zip -d .local && echo "$(pwd)/.local/bin" >> $GITHUB_PATH
```
[the remark this answers is not in the document]
Comment (markcty) [reply .github/workflows/ci.yml:40]: Done. In .github/scirpts/install_deps.sh

File: curp/src/client.rs [hunk 2]
```diff
@@ -128,37 +178,121 @@ where
...
+                let mut state = self.state.write();
+                debug!("Fetch leader succeeded, leader set to {}", leader);
+                state.term = max_term;
+                state.set_leader(leader);
+                return;
+            }
+
+            // wait until the election is completed
+            // TODO: let user configure it according to average leader election cost
+            tokio::time::sleep(Duration::from_micros(500)).await;
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/client.rs]: Done.

File: curp/src/client.rs [hunk 3]
```diff
@@ -128,37 +178,121 @@ where
...
+                        max_term = resp.term;
+                        leader = Some(leader_id);
+                    }
+                }
+                if resp.term > max_term {
+                    max_term = resp.term;
+                }
+            }
+
+            if let Some(leader) = leader {
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/client.rs:301]: Fixed.

File: curp/src/client.rs [hunk 4]
```diff
@@ -87,16 +129,36 @@ where
...
+                        // reset term only when the resp has leader id to prevent:
+                        // If a server loses contact with its leader, it will update its term for election. Since other servers are all right, the election will not succeed.
+                        // But if the client learns about the new term and updates its term to it, it will never get the true leader.
+                        if let Some(ref leader_id) = resp.leader_id {
+                            ok_cnt = 0;
+                            state.update_to_term(resp.term());
+                            state.set_leader(leader_id.clone());
+                            execute_result = None;
+                        }
+                        true
```
Comment (rogercloud) [inline curp/src/client.rs:145]: I guess we can't count it as valid term unless the leader_id is determined.

**Review by rogercloud — COMMENTED** [review]
Why the test utils are not in the test directory? Are they UT utils?

File: curp/src/channel/key_spmc.rs [hunk 5]
```diff
@@ -294,8 +294,8 @@ mod test {
         });
     }
 
-    // Test the receiver should not block the thread. A bug was found that the receiver might block the whole tokio worker thread when recv().await is called. This test verifies that it was fixed.
-    // The test will be blocked should the async_recv() in SpmcKeyBasedReceiver::recv() is changed to recv().
+    // Test the receiver should not block the thread. A bug was found that the receiver might block the whole tokio worker thread when recv().await is called. This tests verifies that it was fixed.
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/channel/key_spmc.rs]: I guess this is accidentally modified in some refactoring. Anyway, this file has been removed.

File: curp/Cargo.toml [hunk 6]
```diff
@@ -37,6 +37,9 @@ tracing-opentelemetry = "0.18.0"
 [dev-dependencies]
 itertools = "0.10.3"
 tracing-subscriber = { version = "0.3.16", features = ["env-filter"] }
+tracing-test = "0.2.3"
+tower = { version = "0.4.13", features = ["filter"] }
+anyhow = "1.0.66"
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/Cargo.toml]: Removed.

**Comment by markcty** [comment]
> 

Yes, they are used in UT tests.

**Comment by markcty** [comment]
> Why the test utils are not in the test directory? Are they UT utils?

Yes, they are used in UT tests.

File: curp/src/client.rs (hunk 4 above)
Comment (markcty) [reply curp/src/client.rs]: Discussed in Feishu.

**Review by rogercloud — APPROVED** [review]

**Pull request merged**